Limit concurrent uploads to 20 to prevent memory exhaustion on huge websites

FraserThompson opened the pull request:
- Uploads are added to a queue and limited to 20 at a time (this should probably be a configurable parameter)
- Upload progress is displayed for each item

JoshuaWalsh commented:
Fixes #57.

JoshuaWalsh commented on src/bin.ts:
```diff
@@ -91,6 +103,56 @@ const createSafeS3Key = (key: string): string => {
+    const hashStream = stream.pipe(createHash('md5').setEncoding('hex'));
+
+    streamToPromise(hashStream)
+        .then((data) => {
```
I'd really like to see all usage of .then converted into awaits. It makes the code a lot more readable.

JoshuaWalsh commented on src/bin.ts:
```diff
@@ -91,6 +103,56 @@ const createSafeS3Key = (key: string): string => {
     return key;
 };
 
+const uploadQueue = queue((task: UploadTask, callback: (key?: string, err?: any) => void) => {
```
I think uploadQueue should be created/scoped within the deploy function, since it is populated within that function and is drained within that function.

JoshuaWalsh commented on src/bin.ts:
```diff
@@ -178,75 +240,52 @@ const deploy = async ({ yes, bucket }: { yes: boolean, bucket: string }) => {
-                    }
-                }).promise();
+        
+        uploadQueue.drain = async () => {
```
Not a fan of using a callback here. The cleanest option would be to use parallelLimit instead, as it [supports promises/awaiting](https://github.com/caolan/async/blob/master/test/es2017/awaitableFunctions.js).

Another idea would be to make a helper method to convert drain into an awaitable, something like: (untested)

```typescript
const drainQueue(queue: QueueObject): Promise<void> => {
    return new Promise((resolve, reject) => {
        queue.drain = resolve;
        queue.error = (err, task) => {
            queue.kill();
            reject(err); // If you want to get really fancy, make a custom error object that includes task in it
        };
    });
}
```

JoshuaWalsh commented on typings/asyncQeueu.ts:
```diff
@@ -0,0 +1 @@
+declare module 'async/queue';
```
This filename is spelled incorrectly, it should be asyncQueue.ts

Btw, why is this necessary? Could we remove this file and in bin.ts do:

```typescript
import {queue} from 'async';
```

FraserThompson commented:
Sorry it took so long. I've switched to using parallelLimit and cleaned up all of those issues except one...

As far as I can tell parallelLimit doesn't supporting awaiting... Unless I'm doing it wrong. With just await it still doesn't seem to wait until all of the functions in the array are called.

So it's still using a callback for when it's done everything. Which obviously isn't great but I'm not sure what else to do...

JoshuaWalsh commented on src/bin.ts:
```diff
@@ -180,75 +184,83 @@ const deploy = async ({ yes, bucket }: { yes: boolean, bucket: string }) => {
-                        ACL: config.acl === null ? undefined : (config.acl || 'public-read'),
-                        ContentType: mime.getType(path) || 'application/octet-stream',
-                        ...getParams(key, params)
+            uploadQueue.push(async (callback: () => void) => {
```
Having an ES2017 async function with a Node-style callback argument is a bit confusing. According to the [async docs](https://caolan.github.io/async/global.html) you don't need to use the callback argument, you can just return the value like normal in the async function. If you tried this and it didn't work, it's most likely because we currently transpile our async/await functions due to #60.

Until we act on #60, I suggest you use [asyncify](https://caolan.github.io/async/docs.html#asyncify) to work around this.

JoshuaWalsh commented on src/bin.ts:
```diff
@@ -95,6 +96,8 @@ const createSafeS3Key = (key: string): string => {
 const deploy = async ({ yes, bucket }: { yes: boolean, bucket: string }) => {
     const spinner = ora({ text: 'Retrieving bucket info...', color: 'magenta' }).start();
+    
+    const uploadQueue: any[] = [];
```
I think this should be typed as `Array<() => Promise<void>>` or `(() => Promise<void>)[]`.

FraserThompson commented:
Aha! I was really confused why await wasn't working for async.parallellLimit and why it seemed to need a callback.

No more callbacks now. One weird thing is the promisified parallelLimit seems to not like the type on the array it gets so I'm casting it to any... Might be something better to do here.

Worth noting that this also removes the `process.exit(1)` on upload failure. I found this more annoying that anything since it meant I could walk away from a large upload and then come back to find it'd failed on a file for some reason and stopped the whole thing. It still logs failures so the user still knows about them... But there might be a good reason for it I haven't considered.

JoshuaWalsh commented:
> No more callbacks now. One weird thing is the promisified parallelLimit seems to not like the type on the array it gets so I'm casting it to any... Might be something better to do here.

I think because we're using asyncify the type of the array should be Array<AsyncFunction<void, Error>>. [See here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/695b694eb95c6c056b88f152e0949edf2ea09bcc/types/async/index.d.ts#L175). But I'm not 100% sure if this will fix the issue, because apparently asyncify returns `any`, [see here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/695b694eb95c6c056b88f152e0949edf2ea09bcc/types/async/index.d.ts#L245). Maybe worth a try? If we do need to keep the cast, can you please put a comment explaining why we do that typecast?

> Worth noting that this also removes the `process.exit(1)` on upload failure. I found this more annoying that anything since it meant I could walk away from a large upload and then come back to find it'd failed on a file for some reason and stopped the whole thing. It still logs failures so the user still knows about them... But there might be a good reason for it I haven't considered.

I think that this should be done as a separate issue/PR. People may be relying on the current behaviour, especially in CI environments. It's a breaking change and we'd have to consider it properly.

FraserThompson commented:
Fair enough, I think error handling in general could probably be improved in other ways too (like retrying on error) so it should definitely be another issue.

The un-promisified `parallelLimit` is happy with an array of type `Array<AsyncFunction<void, Error>>` but apparently `util.promisify(parallelLimit)` expects a `Dictionary<AsyncFunction<{}, {}>>` and I have no idea why.

JoshuaWalsh commented:
> The un-promisified `parallelLimit` is happy with an array of type `Array<AsyncFunction<void, Error>>` but apparently `util.promisify(parallelLimit)` expects a `Dictionary<AsyncFunction<{}, {}>>` and I have no idea why.

It seems to be because the parallelLimit function is overloaded and util.promisify's TS definition has a bug that causes the wrong overload to be selected. DefinitelyTyped/DefinitelyTyped#20497.

I've made a small code cleanliness commit, hope you don't mind. I moved the promisifedParallelLimit function to the top of the file, as it's a utility method that we could potentially use elsewhere.

JoshuaWalsh commented:
Alright, it's merged. Thanks for all your hard work on this @FraserThompson!

iamkhalidbashir commented on src/bin.ts:
```diff
@@ -180,52 +187,57 @@ const deploy = async ({ yes, bucket }: { yes: boolean, bucket: string }) => {
         await streamToPromise(stream as any as Readable); // todo: find out why the typing won't allow this as-is
-        await Promise.all(promises);
-
+        await promisifiedParallelLimit(uploadQueue, 20);
```
This limit should be made a config option. As the deployment breaks on large sites uploading with slow internet.
